### ISSUE TITLE
`mypy` and `ruff` do not detect as many errors when run through `pre-commit`

### ISSUE THREAD

**Issue by jeremyestein** [UCLH-Foundry/PIXL#305]
### Definition of Done / Acceptance Criteria

- [x] Mypy will detect common typing errors to an appropriate level as determined by the dev team (it is highly configurable)
- [x] Ditto ruff
- [x] Satisfactorily investigate why settings are different when run directly vs via pre-commit

### Details and Comments

I don't ever recall `mypy` failing in our pre-commit checks - is it just turned down to a very low level? Even the most obvious of typing errors isn't detected:
Quick test: add line `upload_parquet_files(42)` to `pixl_core/src/core/upload.py`.
Running `mypy pixl_core/src/core/upload.py` gives `error: Argument 1 to "upload_parquet_files" has incompatible type "int"; expected "ParquetExport"`. My IDE also gives the error.
But running `pre-commit run --all-files --verbose` passes just fine. I would expect it to fail too.

**Comment by jeremyestein** [comment]
The problem seems to exist for `ruff` too.
Compare pre-commit:
```
$  pre-commit run --all-files --verbose
...
ruff.....................................................................Passed
- hook id: ruff
- duration: 0.03s
...
```

with running ruff directly:
```
$ ruff .
...
pixl_ehr/src/pixl_ehr/main.py:158:10: ASYNC101 Async functions should not call `open`, `time.sleep`, or `subprocess` methods
pixl_ehr/tests/test_processing.py:40:20: LOG002 Use `__name__` with `logging.getLogger()`
Found 2 errors.
...
```
Neither `LOG002` nor `ASYNC101` are explicitly mentioned in any config file.

**Comment by milanmlft** [comment]
For `ruff` it's probably because we're using a hopelessly outdated version in pre-commit. `mypy` not sure yet, but likely something similar. Will open a PR soonish to update the `ruff` stuff and see if I can find a fix for `mypy`.

**Comment by milanmlft** [comment]
OK so I got `ruff` fixed in #347 and did some updates for `mypy` as well.
BUT, `mypy` is still not catching much. E.g. the example of Jeremy still passes (while it shouldn't).
